### PR TITLE
Add HTTP-level tests for KV subkeys endpoint

### PR DESCRIPTION
**Pull request by ccapurso**
Adding HTTP-level tests to ensure that KVv2 subkeys endpoint is responding with the correct status codes. Similar unit tests exist at the logical-level in the `vault-plugins-secret-kv` repository.

### CODE REVIEW

**Review by akshya96 — APPROVED** [review]
Lgtm

File: vault/external_tests/kv/kv_subkeys_test.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+		t.Fatalf("expected subkeys request to fail with %d status code, resp: %#v", http.StatusNotFound, apiResp)
+	}
+
+	secret, err := api.ParseSecret(apiResp.Body)
+	if err != nil {
+		t.Fatalf("failed to parse resp body, err: %v", err)
+	}
+
+	if subkeys, ok := secret.Data["subkeys"]; !ok || subkeys != nil {
+		t.Fatalf("expected nil subkeys, actual: %#v", subkeys)
```
Comment (pmmukh) [inline vault/external_tests/kv/kv_subkeys_test.go:120]: is the !ok check valid for this message? I'm kinda thinking that might be a separate message with a message like `key "subkeys" not found in response`

File: vault/external_tests/kv/kv_subkeys_test.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	if apiResp.StatusCode != http.StatusNotFound {
+		t.Fatalf("expected subkeys request to fail with %d status code, resp: %#v", http.StatusNotFound, apiResp)
+	}
+
+	secret, err := api.ParseSecret(apiResp.Body)
+	if err != nil {
+		t.Fatalf("failed to parse resp body, err: %v", err)
+	}
+
+	if subkeys, ok := secret.Data["subkeys"]; !ok || subkeys != nil {
```
Comment (pmmukh) [inline vault/external_tests/kv/kv_subkeys_test.go:201]: similar nit as above

File: vault/external_tests/kv/kv_subkeys_test.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+	}
+
+	resp, err = c.Logical().Write("kv/destroy/foo", destroyVersions)
+	if err != nil {
+		t.Fatalf("destroy failed, err :%v, resp: %#v\n", err, resp)
+	}
+
+	req := c.NewRequest("GET", "/v1/kv/subkeys/foo")
+	apiResp, err := c.RawRequestWithContext(context.Background(), req)
+	if resp != nil {
```
Comment (pmmukh) [inline vault/external_tests/kv/kv_subkeys_test.go:184]: hmm wanted to ask cause not clear on this, why is `resp` nil-checked before closing the apiResp body?

File: vault/external_tests/kv/kv_subkeys_test.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+	if !ok {
+		t.Fatalf("metadata not present in response or invalid, metadata: %#v", secret.Data["metadata"])
+	}
+
+	if destroyed, ok := metadata["destroyed"].(bool); !ok || !destroyed {
+		t.Fatalf("expected destroyed to be true, metadata: %#v", metadata)
+	}
+}
+
+func TestKV_Subkeys_CurrentVersion(t *testing.T) {
```
Comment (pmmukh) [inline vault/external_tests/kv/kv_subkeys_test.go:216]: i know this is how the unit test also works so might just be rehashing old convos, but for checking that without a version number the latest is returned, should we have 2 versions of the data here ?

**Review by pmmukh — COMMENTED** [review]
some small nits, besides which i have one doubt, should there also be tests here for checking the depth param like the unit tests do ?

File: vault/external_tests/kv/kv_subkeys_test.go (hunk 1 above)
Comment (ccapurso) [reply vault/external_tests/kv/kv_subkeys_test.go:131]: https://github.com/hashicorp/vault/pull/14079/commits/5db99a2940f1a62821de002de02e6cb70aec47c0

File: vault/external_tests/kv/kv_subkeys_test.go (hunk 2 above)
Comment (ccapurso) [reply vault/external_tests/kv/kv_subkeys_test.go]: https://github.com/hashicorp/vault/pull/14079/commits/5db99a2940f1a62821de002de02e6cb70aec47c0

File: vault/external_tests/kv/kv_subkeys_test.go (hunk 4 above)
Comment (ccapurso) [reply vault/external_tests/kv/kv_subkeys_test.go:239]: https://github.com/hashicorp/vault/pull/14079/commits/5db99a2940f1a62821de002de02e6cb70aec47c0

**Comment by ccapurso** [comment]
> some small nits, besides which i have one doubt, should there also be tests here for checking the depth param like the unit tests do ?

I have added these tests mainly to ensure that the HTTP status codes were appropriately set as there are certain cases that are important. My goal wasn't necessarily to exhaustively test the underlying functionality of the endpoint.

**Review by pmmukh — APPROVED** [review]

**Comment by pmmukh** [comment]
> I have added these tests mainly to ensure that the HTTP status codes were appropriately set as there are certain cases that are important. My goal wasn't necessarily to exhaustively test the underlying functionality of the endpoint.

ah gotcha, makes sense!!

**Pull request merged**